include/seqan3/io/concept/stream.hpp

```diff
@@ -0,0 +1,274 @@
+     using off_type    = typename stream_::off_type;
+     using pos_type    = typename stream_::pos_type;
+ };
+
+ template <typename char_t, typename traits_t>
+ struct stream_traits<std::basic_ofstream<char_t, traits_t>>
+ {
+     using stream_ = std::basic_ofstream<char_t, traits_t>;
+
+     using char_type   = typename stream_::char_type;
+     using traits_type = typename stream_::traits_type;
+     using int_type    = typename stream_::int_type;
+     using off_type    = typename stream_::off_type;
+     using pos_type    = typename stream_::pos_type;
+ };
+
+ template <typename char_t, typename traits_t>
+ struct stream_traits<std::basic_ostringstream<char_t, traits_t>>
+ {
+     using stream_ = std::basic_ostringstream<char_t, traits_t>;
+     using char_type   = typename stream_::char_type;
+     using traits_type = typename stream_::traits_type;
+     using int_type    = typename stream_::int_type;
+     using off_type    = typename stream_::off_type;
+     using pos_type    = typename stream_::pos_type;
+ };
+ }  // namespace seqan3::detail
+
+namespace seqan3
+{
+
+template <typename stream_t>
+struct char_type
+{
+    using type = typename detail::stream_traits<std::decay_t<stream_t>>::char_type;
+};
+
+template <typename stream_t>
+using char_type_t =  typename char_type<stream_t>::type;
+
+template <typename stream_t>
+struct traits_type
+{
+    using type = typename detail::stream_traits<std::decay_t<stream_t>>::traits_type;
+};
+
+template <typename stream_t>
+using traits_type_t =  typename traits_type<stream_t>::type;
+
+template <typename stream_t>
+struct int_type
+{
+    using type = typename detail::stream_traits<std::decay_t<stream_t>>::int_type;
+};
+
+template <typename stream_t>
+using int_type_t =  typename int_type<stream_t>::type;
+
+template <typename stream_t>
+struct off_type
+{
+    using type = typename detail::stream_traits<std::decay_t<stream_t>>::off_type;
+};
+
+template <typename stream_t>
+using off_type_t =  typename off_type<stream_t>::type;
+
+template <typename stream_t>
+struct pos_type
+{
+    using type = typename detail::stream_traits<std::decay_t<stream_t>>::pos_type;
+};
+
+template <typename stream_t>
+using pos_type_t =  typename pos_type<stream_t>::type;
+
+template <typename t>
+concept bool input_stream_concept =
+    requires(t stream, typename t::char_type val)
+{
+    typename char_type<t>::type;
+    typename traits_type<t>::type;
+    typename int_type<t>::type;
+    typename pos_type<t>::type;
+    typename off_type<t>::type;
+
+    // formatted input.
+    { operator>>(stream, val) } -> auto&;
+};
+
+template <typename t>
+concept bool output_stream_concept =
+    requires(t stream)
+{
+    typename char_type<t>::type;
+    typename traits_type<t>::type;
+    typename int_type<t>::type;
+    typename pos_type<t>::type;
+    typename off_type<t>::type;
+
+    // formatted output.
+    { operator<<(stream, typename t::char_type{}) } -> auto&;
+};
```

Comment:
First you define structs with members, than you define free metafunctions that retrieve these members, than you define concepts that require these metafunctions... and this all for some types that seem to be the same for every type? Seems like needless over-generalization to me, but maybe we can discuss why its required.